concurrency/limited_concurrency_singleflight.go

```diff
@@ -0,0 +1,101 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+// LimitedConcurrencySingleFlight ensures that across all concurrent calls to ForEachNotInFlight, only up to maxConcurrent
+// tokens are running concurrently. See the docs of ForEachNotInFlight for the uniqueness semantics of tokens.
+//
+// An example use case for LimitedConcurrencySingleFlight is to run a periodic job concurrently for N users. Sometimes a single user's
+// job may take longer than the period. And this will block future scheduled jobs until the first job for that user completes.
+// Call LimitedConcurrencySingleFlight.ForEachNotInFlight in its own goroutine with user IDs as tokens.
+// This will run the jobs concurrently and make sure that two jobs for the same user don't run concurrently.
+type LimitedConcurrencySingleFlight struct {
+	inflightTokensMx sync.Mutex
+	inflightTokens   map[string]struct{}
+	inflightCalls    sync.WaitGroup
+	semaphore        chan struct{}
+}
+
+func NewWorkerPool(maxConcurrent int) *LimitedConcurrencySingleFlight {
```

Comment:
The name of the constructor is now completely different from the type which it constructs.

Reply:
aii. not merged to main and it's already outdated. thanks for spotting